xapian-bindings/java/java.i

```diff
@@ -174,8 +174,11 @@ namespace Xapian {
     bool hasNext() const { return Xapian::iterator_valid(*self); }
 }
 
+%ignore MSetIterator::next();
+%javamethodmodifiers MSetIterator::next(int) "private"
+
 %extend MSetIterator {
-    Xapian::docid next () {
+    Xapian::docid next (int pseudoArg) {
```

Comment:
You don't really want the existing `next()` method's current functionality anyway, as it dereferences the iterator before advancing it and returns that value, and you aren't interested in that.

But the current iterator wrappers weren't present in the hand-coded JNI Java bindings in 1.2.x, so compatibility with them isn't a big concern - it's fine to throw out the current `next()` behaviour to implement something smarter.

So I'd just replace the `next()` added by extend with a simpler helper method which does what you want.

Ideally it would check `Xapian::iterator_valid(*self)` as well as incrementing the iterator, but unfortunately that's too late as you already read values in the `MatchDescriptor` constructor. Really you want (in pseudo-code):

```
if (!iterator_valid()) {
    return null;
}

MatchDescriptor descriptor = new MatchDescriptor();
_next();
return descriptor;
```

It would probably work to use `%extend` to add an `iterator_valid()` method to the wrapped object.

There's probably scope for optimising this - `next()` then makes 7 JNI calls when there's a next element, and we could potentially just make one to check validity, read docid, rank, percent, etc, increment the iterator and return all the values.  But getting this working and done for all the iterators is most important - optimising the implementation can be done later (and profiling it to see where the time is actually spent would be wise first anyway).